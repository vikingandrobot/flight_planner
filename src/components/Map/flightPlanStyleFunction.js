import { Stroke, Style, Fill } from 'ol/style';
import { Circle } from 'ol/geom';

// Received the vector feature and returns the corresponding styles.
export default function flightPlanStyleFunction(feature) {
  const styles = [
    // linestring
    new Style({
      stroke: new Stroke({
        color: '#31a4eb',
        width: 3,
      }),
      zIndex: 0,
    }),
  ];

  const coordinates = feature.getGeometry().getCoordinates();

  return styles.concat(coordinates.map(coordinate => new Style({
    geometry: new Circle(coordinate, 4),
    fill: new Fill({
      color: '#6bc7ff',
    }),
    stroke: new Stroke({
      color: '#31a4eb',
      width: 2,
    }),
    zIndex: 1,
  })));
}
