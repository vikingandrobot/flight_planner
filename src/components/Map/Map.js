import React, { useRef, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import 'ol/ol.css';
import OlMap from 'ol/Map';
import View from 'ol/View';
import TileLayer from 'ol/layer/Tile';
import VectorLayer from 'ol/layer/Vector';
import Feature from 'ol/Feature';
import LineString from 'ol/geom/LineString';
import OSM from 'ol/source/OSM';
import VectorSource from 'ol/source/Vector';
import ResizeObserver from 'resize-observer-polyfill';

import flightPlanStyleFunction from './flightPlanStyleFunction';

import styles from './Map.module.scss';

function Map({
  className = '',
  center = [0, 0],
  zoom = 1,
  onClick,
  fitViewToFlightPlan,
  flightPlan = null,
}) {
  const targetRef = useRef(null);
  const mapRef = useRef(null);
  const viewRef = useRef(null);
  const flightPlanSourceRef = useRef(new VectorSource());
  const sizeObserverRef = useRef();

  const hasFlightPlan = !!flightPlan;

  const handleClick = useCallback((e) => {
    if (onClick) {
      onClick(e.coordinate);
    }
  }, [onClick]);

  useEffect(() => {
    function createMap() {
      if (targetRef.current && !mapRef.current) {
        viewRef.current = new View({ center, zoom });

        mapRef.current = new OlMap({
          view: viewRef.current,
          layers: [
            new TileLayer({
              source: new OSM(),
            }),
            new VectorLayer({
              source: flightPlanSourceRef.current,
              style: flightPlanStyleFunction,
            })
          ],
          target: targetRef.current,
        });

        // Size observer for the map
        sizeObserverRef.current = new ResizeObserver(() => {
          mapRef.current.updateSize()
        });
        sizeObserverRef.current.observe(targetRef.current);
      }
    }

    createMap();

    return () => {
      // Clean up
      if (sizeObserverRef.current) {
        sizeObserverRef.current.disconnect();
      }
    }
  }, []);

  useEffect(() => {
    if (mapRef.current && onClick) {
      mapRef.current.un('click', handleClick);
      mapRef.current.on('click', handleClick);
    }

    return () => {
      mapRef.current.un('click', handleClick);
    }
  }, [onClick, handleClick]);

  useEffect(() => {
    flightPlanSourceRef.current.clear();
    if (flightPlan) {
      const feature = new Feature({
        geometry: new LineString(flightPlan.points),
      });
      flightPlanSourceRef.current.addFeature(feature);

      if (fitViewToFlightPlan && viewRef.current) {
        viewRef.current.fit(feature.getGeometry(), {padding: [100, 100, 100, 100]});
      }
    }
  }, [hasFlightPlan, JSON.stringify(flightPlan?.points)])

  return (
    <div ref={targetRef} className={`${styles.mapContainer} ${className}`} />
  )
}

Map.propTypes = {
  // A class name to pass to the map container to modify its style
  className: PropTypes.string,
  // Center of the map
  center: PropTypes.arrayOf(PropTypes.number),
  zoom: PropTypes.number,
  fitViewToFlightPlan: PropTypes.bool,
  flightPlan: PropTypes.shape({
    points: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
  })
};

export default Map;
