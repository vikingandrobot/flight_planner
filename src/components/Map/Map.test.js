import { render } from '@testing-library/react';

import * as OlMap from 'ol/Map';
import * as OlView from 'ol/View';

import Map from './Map';

jest.mock('ol/Map');
jest.mock('ol/View');
jest.mock('ol/layer/Tile');
jest.mock('ol/layer/Vector');
jest.mock('ol/Feature');
jest.mock('ol/geom');
jest.mock('ol/source/OSM');
jest.mock('ol/source/Vector');

describe('Map', () => {

  test('should pass down the class name to the container', () => {
    const {Â container } = render(<Map className='test-classname' />);
    expect(container).toMatchSnapshot();
  })

  test('should create a new Open Layers map on mount', () => {
    const viewOptions = {
      center: [0, 0],
      zoom: 11,
    };

    render(<Map center={viewOptions.center} zoom={viewOptions.zoom} />);

    expect(OlMap.default).toHaveBeenCalledTimes(1);
    expect(OlView.default).toHaveBeenCalledWith(viewOptions);
  });
});
