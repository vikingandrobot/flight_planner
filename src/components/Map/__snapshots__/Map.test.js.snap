// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map should pass down the class name to the container 1`] = `
<div>
  <div
    class="mapContainer test-classname"
  />
</div>
`;
